### PR TITLE
Add unified arch for gemma4

### PR DESCRIPTION
**Pull request by neilmehta24**
Summary of changes:
- Removed `_BatchedLogitsProcessorAdapter` now that https://github.com/ml-explore/mlx-lm/pull/1115 has landed
- Add vision addon for gemma4
  - The vision features for gemma4 include both embeddings and per-layer-inputs. So, add a model patch to ensure that `per_layer_inputs` is wired into the language model.
  - Unapply `embed_scale` before passing into the language model, because mlx-lm will apply it.
  - Add tests for vision input, text input, and prompt caching.

To verify that this implementation is correct, I ensured the following things:
- The `input_ids` reaching the first `mlx-lm` pass match native `mlx-vlm`
- The `per_layer_inputs` reaching that same pass also match native `mlx-vlm`
- The embeddings seen by `mlx-lm` match native behavior. The only difference is a slight rounding error for image embeddings from the un-application of the embed_scale followed by a later re-application.
- I also ran an e2e generation with (1) a multi-image summary task and (2) a single-image transcription task. I didn't notice any quality degradation between mlx-vlm and this unified impl.

Codex review:
Screenshot 2026-04-07 at 12 27 01 PM

### CODE REVIEW

File: mlx_engine/model_kit/patches/gemma4.py
Comment (will-lms) [inline mlx_engine/model_kit/patches/gemma4.py:1]: Did you confirm the logits are the same for both text-only and vision prompts before and after the patch? Consider adding tests to `test_patched_models.py` in line with the Qwen 3.5 heavy tests to verify.

Comment (neilmehta24) [reply mlx_engine/model_kit/patches/gemma4.py:1]: I verified that the logits matched for text-only work, and that the logits are close-enough within a tolerance for image+text work. I added a test for each of these cases.

File: tests/patched_model_test_utils.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    pytest.param(
+        "lmstudio-community/Qwen3.5-35B-A3B-MLX-4bit",
+        marks=pytest.mark.heavy,
+        id="moe",
+    ),
+]
+GEMMA4_MODEL_NAME = "lmstudio-community/gemma-4-E2B-it-MLX-4bit"
+GEMMA4_IMAGE_TOPK = 5
+GEMMA4_IMAGE_TOPK_PROB_RTOL = 0.25
+GEMMA4_IMAGE_TOPK_PROB_REF_FLOOR = 1e-3
```
Comment (will-lms) [inline tests/patched_model_test_utils.py]: Feels like these should be in the model-specific test files, not the shared utils.

File: tests/patched_model_test_utils.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    _assert_restorable_binding(
+        OriginalGemma4TextModel,
+        gemma4_text_module.Gemma4TextModel,
+        "Gemma4TextModel",
+    )
+    return _load_unpatched_mlx_lm(
+        model_path,
+        module=gemma4_text_module,
+        replacements={"Gemma4TextModel": OriginalGemma4TextModel},
+    )
```
Comment (will-lms) [inline tests/patched_model_test_utils.py]: Same with these

File: tests/patched_model_test_utils.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+def load_vlm(model_path: Path):
+    result = vlm_load_model(model_path)
+    return result[0] if isinstance(result, tuple) else result
+
+
+def load_vlm_processor(model_path: Path):
+    return load_processor(model_path, add_detokenizer=True)
+
+
+def build_gemma4_prompt(
```
Comment (will-lms) [inline tests/patched_model_test_utils.py]: Same with this

File: tests/test_patched_gemma4.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    relative_differences,
+    resolve_image_token_index,
+    softmax_probabilities,
+    tokenize_prompt,
+    topk_token_ids,
+)
+from tests.shared import read_image_b64
+
+
+def test_gemma4_text_only_generation_patched_matches_unpatched():
```
Comment (will-lms) [inline tests/test_patched_gemma4.py:165]: Should this be heavy? I see the VLM test is.

Comment (neilmehta24) [reply tests/test_patched_gemma4.py:165]: Yes, missed it.  Added `pytestmark = pytest.mark.heavy`

File: tests/patched_model_test_utils.py (hunk 1 above)
Comment (neilmehta24) [reply tests/patched_model_test_utils.py]: Moved these to the model-specific files

**Review by will-lms — APPROVED** [review]
